p2p/src/peer_manager/mod.rs

```diff
@@ -1574,16 +1654,21 @@ where
             let last_heartbeat_max =
                 (last_heartbeat + PEER_MGR_HEARTBEAT_INTERVAL_MAX).expect("Cannot happen 2");
 
+            let time_since_last_chainstate_tip =
+                (now - self.last_chainstate_tip_block_time).unwrap_or(Duration::ZERO);
+            let tip_is_stale = time_since_last_chainstate_tip > stale_tip_time_diff;
+
+            if tip_is_stale {
+                heartbeat_call_needed = true;
+            }
```

Comment:
I'm heavily opposed to solving problems by just adding a bunch of lines. If a bunch of lines are isolated in logic, they should go to another "thing", be it a function, struct or whatever. Please isolate the detection of stale tip. There's no reason why the logic of detecting a tip should be intermixed with other things, like clock and heartbeat stuff.